docs/how-tos/how-to-customize-core-behavior-in-state-machines.mdx

```diff
@@ -0,0 +1,52 @@
+---
+sidebar_position: 2
+---
+
+# How to customize core behavior in state machines
+
+Sometimes you might need to tweak Solidus' core model to fit your business
+needs. In that case, you might want to tweak a state machine to obey your
+extended domain.
+
+Say that you must store the time when a payment has been marked as completed.
+First, you need to add a new `#completed_at` field to the payments table:
+
+```bash
+bin/rails g migration AddCompletedAtToSpreePayments completed_at:time
+bin/rails db:migrate
+```
+
+Next, you can leverage the payment state machine to fill it. You can add an
+`after_transition hook` using an [override][customize-core] (don't forget the setup code in `config/application.rb`):
+
+```ruby title="app/overrides/my_store/payment_set_completed_at.rb"
+# frozen_string_literal: true
+
+module MyStore
+  module PaymentSetCompletedAt
+    def self.prepended(base)
+      base.state_machine.after_transition(to: :completed) do
+        self.completed_at = Time.zone.now
+      end
+    end
+
+    ::Spree::Payment.prepend self
+  end
+end
+```
+
+That's all you need to do. From this moment, the state machine will be
+responsible for storing when a payment is completed.
+
+:::danger
+
+Be aware of not overusing transition hooks in the state machines. When the involved logic requires
+reaching external services or, more generally, is decoupled from the main flow, you're better
+off [leveraging the event bus][how-to-add-orthogonal-behavior]. Otherwise, you
+will eventually run into the typical gotchas and downsides of abusing
+`ActiveRecord` callbacks.
```

Comment:
At this point, we could also point out that the same result could be accomplished following the `How to use custom state machines` guide but replacing the state machine and only adding the `after_transition`.  Maybe a link to the `Customizing states machines` section of the main page could be useful to point people to read the explanation of which method is best for different scenarios.

Reply:
I'm not sure I follow. Wouldn't it be overkilling to replace the whole state machine only to add an `after_transition` hook? I see the replacement as a last resort when you need a lot of flexibility. We're already linking to the guide about adding orthogonal behavior, which I think will be the most common gotcha.

Reply:
I think this will be addressed when we will resolve https://github.com/solidusio/edgeguides/pull/65#discussion_r976447122. Please, ignore it 😄